core/engine/src/module/mod.rs

```diff
@@ -10,11 +10,24 @@
 //! - Link its dependencies together using [`Module::link`].
 //! - Evaluate the module and its dependencies using [`Module::evaluate`].
 //!
-//! The [`ModuleLoader`] trait allows customizing the "load" step on the lifecycle
+//! ### `ModuleLoader` and `RefModuleLoader`
+//!
+//! The [`ModuleLoader`] and [`RefModuleLoader`] traits allow customizing the "load" step on the lifecycle
 //! of a module, which allows doing things like fetching modules from urls, having multiple
 //! "modpaths" from where to import modules, or using Rust futures to avoid blocking the main thread
 //! on loads.
 //!
+//! A type can only implement one of the two traits, since implementing `RefModuleLoader`
+//! will automatically implement `ModuleLoader`, and viceversa. Depending on your use case,
+//! you would want to implement one or the other:
+//!
+//! - `ModuleLoader` takes `Rc<Self>` as receiver, which makes it better for loaders that need to
+//!   asynchronously modify their internal state, since it allows moving the loader into the future
+//!   returned by [`ModuleLoader::load_imported_module`].
+//! - `RefModuleLoader` takes `&Self` as receiver, which disallows moving the loader into the future
+//!   returned by [`RefModuleLoader::load_imported_module`]. This trait is ideal for loaders that mostly
+//!   do everything in a synchronous way, or don't need to modify any internal state.
+//!
```

Comment:
I think, for the sake of saving a lot of questions and maintenance, that we should pick one of those 2 and stick with it. My vote would go for `Rc<Self>`. It's self explanatory, not really unusual for async code, and will be the most future proof one.

I don't think supporting both will do us any favors. The number of times people implement their own `ModuleLoader` is minimal at best anyway (they should use one of our provided implementation).

Reply:
I think this is fair. We can still optimize this in the future with some kind of "projectable" Rc pointer when the Rust team stabilizes custom receivers.

Reply:
Rolled back to the unoptimized but simpler version.